daemon/config/builder.go

```diff
@@ -84,13 +84,17 @@ func (x *BuilderGCFilter) UnmarshalJSON(data []byte) error {
 
 // BuilderGCConfig contains GC config for a buildkit builder
 type BuilderGCConfig struct {
-	Enabled              bool            `json:",omitempty"`
+	Enabled              *bool           `json:",omitempty"`
 	Policy               []BuilderGCRule `json:",omitempty"`
 	DefaultReservedSpace string          `json:",omitempty"`
 	DefaultMaxUsedSpace  string          `json:",omitempty"`
 	DefaultMinFreeSpace  string          `json:",omitempty"`
 }
 
+func (x *BuilderGCConfig) IsEnabled() bool {
+	return x.Enabled == nil || *x.Enabled
+}
+
 func (x *BuilderGCConfig) UnmarshalJSON(data []byte) error {
 	var xx struct {
 		Enabled              bool            `json:",omitempty"`
```

Comment:
I was even considering if we had to modify `BuilderGCConfig. Enabled` to be a pointer, or if we'd only need to modify this temporary struct (unmarshal to xx` and if it's nil, then we set the default to true, which we already do below), but that would require `New()` to set the default (for cases where the `UnmarshalJSON` is not called, and looks like we'd have to patch other places to start with `New()` (default config) instead of starting with an empty config.

Reply:
Modifying the temporary struct doesn't work because `UnmarshalJSON` only gets called _if_ there's a gc section. So doing the change in the temporary struct works for this:

```
"builder": {"gc": {}}
```

But it doesn't work when `"gc"` is missing or if the entire configuration is missing. At the same time, doing it this way ensures that `"enabled": null` isn't accidentally an acceptable value. If the temporary struct has a pointer, then it would be ok for someone to set `null` as the value. If we set `Enabled` to true before running unmarshal then it will be set to true if the attribute isn't present otherwise it'll be overridden and `null` will cause it to error.

Reply:
Yeah, I came to that conclusion as well; I had a bit of a brain-fart and thought it would be called unconditionally, but that's that didn't really make sense;

> for cases where the UnmarshalJSON is not called